drivers/usb/device/usb_dc_mcux_ip3511fs.c

```diff
@@ -0,0 +1,528 @@
+{
+	u8_t ep_abs_idx = EP_ABS_IDX(ep);
+	u8_t *buffer = (u8_t *)s_Device.eps[ep_abs_idx].block.data;
+	u32_t len_to_send;
+
+	if (data_len > s_Device.eps[ep_abs_idx].ep_mps) {
+		len_to_send = s_Device.eps[ep_abs_idx].ep_mps;
+	} else {
+		len_to_send = data_len;
+	}
+	memcpy(buffer, data, len_to_send);
+	s_Device.interface->deviceSend(s_Device.controllerHandle,
+				       ep,
+				       buffer,
+				       len_to_send);
+	if (ret_bytes) {
+		*ret_bytes = len_to_send;
+	}
+	return 0;
+}
+
+int usb_dc_ep_read_wait(u8_t ep, u8_t *data, u32_t max_data_len,
+			u32_t *read_bytes)
+{
+	u8_t ep_idx = EP_ADDR2IDX(ep);
+	u8_t ep_abs_idx = EP_ABS_IDX(ep);
+	u32_t data_len;
+	u8_t *bufp = NULL;
+
+	while (s_Device.eps[ep_abs_idx].ep_occupied) {
+		LOG_ERR("Endpoint is occupied by the controller");
+		return -EBUSY;
+	}
+	if ((ep_idx >= NUM_OF_EP_MAX) || (EP_ADDR2DIR(ep) != USB_EP_DIR_OUT)) {
+		LOG_ERR("Wrong endpoint index/address/direction");
+		return -EINVAL;
+	}
+	/* Allow to read 0 bytes */
+	if (!data && max_data_len) {
+		LOG_ERR("Wrong arguments");
+		return -EINVAL;
+	}
+	/* it is control setup, we should use message.buffer,
+	 * this buffer is from internal setup array.
+	 */
+	bufp = s_Device.eps[ep_abs_idx].transfer_message.buffer;
+	data_len = s_Device.eps[ep_abs_idx].transfer_message.length;
+	if (data_len == USB_UNINITIALIZED_VAL_32) {
+		if (read_bytes) {
+			*read_bytes = 0;
+		}
+		return -EINVAL;
+	}
+	if (!data && !max_data_len) {
+		/* When both buffer and max data to read are zero return the
+		 * available data in buffer.
+		 */
+		if (read_bytes) {
+			*read_bytes = data_len;
+		}
+		return 0;
+	}
+	if (data_len > max_data_len) {
+		LOG_WRN("Not enough room to copy all the data!");
+		data_len = max_data_len;
+	}
+	if (data != NULL) {
+		memcpy(data, bufp, data_len);
+	}
+	if (read_bytes) {
+		*read_bytes = data_len;
+	}
+
+	if (EP_ADDR2IDX(ep) == USB_ENDPOINT_CONTROL) {
+		u8_t isSetup = s_Device.eps[0].transfer_message.isSetup;
+		u8_t *buffer = s_Device.eps[0].transfer_message.buffer;
+
+		if (isSetup) {
+			if (((usb_setup_struct_t *)buffer)->wLength == 0) {
+				s_Device.setupDataStage = SETUP_DATA_STAGE_DONE;
+			} else if (((usb_setup_struct_t *)buffer)->bmRequestType & USB_REQUEST_TYPE_DIR_MASK) {
+				s_Device.setupDataStage = SETUP_DATA_STAGE_IN;
+			} else {
+				s_Device.setupDataStage = SETUP_DATA_STAGE_OUT;
+			}
+		} else {
+			if (s_Device.setupDataStage != SETUP_DATA_STAGE_DONE) {
+				if ((data_len >= max_data_len) || (data_len < s_Device.eps[0].ep_mps)) {
+					s_Device.setupDataStage = SETUP_DATA_STAGE_DONE;
+				}
+			}
+		}
+	}
+	return 0;
+}
+
+int usb_dc_ep_read_continue(u8_t ep)
+{
+	/* select the index of the next endpoint buffer */
+	u8_t ep_abs_idx = EP_ABS_IDX(ep);
+
+	if (s_Device.eps[ep_abs_idx].ep_occupied) {
+		LOG_WRN("endpoint 0x%x already occupied", ep);
+		return -EBUSY;
+	}
+
+	if (EP_ADDR2IDX(ep) == USB_ENDPOINT_CONTROL) {
+		if (s_Device.setupDataStage == SETUP_DATA_STAGE_DONE) {
```

Comment:
What is setupDataStage used for?

Reply:
NXP controller driver has a build-in USB device stack? Please describe it widely.

Reply:
Yes, NXP has USB stack in the MCUXpresso SDK package. I take the part (controller driver) and implement one shime for Zephyr.